File: cpp/12_inheritance.cpp
// ============================================================
// Topic:  Inheritance
// Summary: A subclass reuses a base class, overrides virtual methods, calls base.
// Run:     g++ -std=c++17 cpp/12_inheritance.cpp -o cpp/12_inheritance && cpp/12_inheritance
// ============================================================

#include <iostream>
#include <string>

class Animal {
protected:
    std::string name;

public:
    Animal(const std::string& name) : name(name) {}
    virtual ~Animal() = default;                 // virtual dtor for safe polymorphism

    // `virtual` enables runtime dispatch to overrides
    virtual std::string sound() const { return "some sound"; }
    virtual std::string describe() const { return name + " says " + sound(); }
};

// ----- 1. `: public Animal` inherits; `Animal(name)` calls the base ctor -----
class Dog : public Animal {
    std::string breed;

public:
    Dog(const std::string& name, const std::string& breed)
        : Animal(name), breed(breed) {}

    std::string sound() const override { return "Woof"; }
    std::string describe() const override { return Animal::describe() + " (" + breed + ")"; }
};

// ----- 2. Subclass that only overrides one method -----
class Cat : public Animal {
public:
    Cat(const std::string& name) : Animal(name) {}
    std::string sound() const override { return "Meow"; }
};

int main() {
    std::cout << Animal("Creature").describe() << std::endl;     // Creature says some sound
    std::cout << Dog("Rex", "Labrador").describe() << std::endl; // Rex says Woof (Labrador)
    std::cout << Cat("Whiskers").describe() << std::endl;        // Whiskers says Meow
    return 0;
}

// Cross-language note:
//   C++: methods must be `virtual` to dispatch dynamically; subclasses add `override`.
//   `: public Animal` inherits; `Animal(name)` in the init list chains the base ctor;
//   `Animal::describe()` calls the parent version. Always give a polymorphic base a virtual dtor.

File: rust/12_inheritance.rs
// ============================================================
// Topic:  Inheritance  (Rust has NONE — reuse via traits + composition)
// Summary: A trait with a default method gives shared behavior; types override it.
// Run:     rustc rust/12_inheritance.rs -o rust/12_inheritance && rust/12_inheritance
// ============================================================

// Rust deliberately has no class inheritance. The idiomatic replacements are:
//   - a TRAIT with default methods   -> shared behavior (what a base class gives you)
//   - COMPOSITION (structs holding other structs) -> "is built from", not "is a"

trait Animal {
    fn name(&self) -> &str;                       // required (like an abstract getter)
    fn sound(&self) -> String {                   // default impl = shared behavior
        "some sound".to_string()
    }
    fn describe(&self) -> String {                // default uses the other methods
        format!("{} says {}", self.name(), self.sound())
    }
}

struct GenericAnimal {
    name: String,
}
impl Animal for GenericAnimal {
    fn name(&self) -> &str { &self.name }         // takes both defaults as-is
}

struct Dog {
    name: String,
    breed: String,
}
impl Animal for Dog {
    fn name(&self) -> &str { &self.name }
    fn sound(&self) -> String { "Woof".to_string() }      // override the default
    fn describe(&self) -> String {                         // extend (no `super`)
        format!("{} says {} ({})", self.name, self.sound(), self.breed)
    }
}

struct Cat {
    name: String,
}
impl Animal for Cat {
    fn name(&self) -> &str { &self.name }
    fn sound(&self) -> String { "Meow".to_string() }       // override only sound
}

fn main() {
    let a = GenericAnimal { name: "Creature".to_string() };
    let d = Dog { name: "Rex".to_string(), breed: "Labrador".to_string() };
    let c = Cat { name: "Whiskers".to_string() };
    println!("{}", a.describe()); // Output: Creature says some sound
    println!("{}", d.describe()); // Output: Rex says Woof (Labrador)
    println!("{}", c.describe()); // Output: Whiskers says Meow
}

// Cross-language note:
//   No `extends`, no base class, no `super`. A trait's DEFAULT methods play the role of
//   inherited behavior; each type re-implements only what differs. To reuse *data*, a
//   struct embeds another struct (composition) rather than inheriting fields.
